Add overwrite attribute to tfe_variable

scottwinkler opened the pull request:
## Description

The recent breaking change to the TEF provider severely affected my organization. We are using the TFE provider to manage all of our sensitive variables, across hundreds of workspaces. When the `tfe_variable` resource was changed to reflect changes in the new API, before the patch was issued, Read() removed every single variable resource from the state file. Now we have close to a 1000 variable resources that are unmanaged. Currently, if I want to bring these under Terraform management I have to manually delete each and every single resource otherwise there is a collision about keys already being taken, or import them one at a time. 

This change proposes adding a new attribute called "overwrite" which will overwrite any existing variable and assume its variable ID, rather than throwing an error. These means that my previous problem of having ~1000 unmanaged variable resources can be fixed with a simple `terraform apply`. Furthermore, this fixes an annoying problem where people create variables in the UI with the same key name (for example for access keys) which then causes my Terraform configuration to fail. Overwrite means that I will overwrite whatever previous value was set, because I don't care what it is. This is an optional attribute, with the default value set to false. Leaving it as the default means that the behavior doesn't change from what it is currently.

## Example

If overwrite is left to its default value of false then attempting to create a variable resource with the same key as an existing one will result in an error same as before:

```
resource "tfe_variable" "var" {
  key          = "my_key"
  value        = "my_value"
  category     = "terraform"
  workspace_id = var.workspace_id
}
```

```
OSXSWINKMBP15:tmp swinkler$ terraform apply -auto-approve
tfe_variable.var: Creating...

Error: Error creating terraform variable my_key: variable already exists

  on main.tf line 10, in resource "tfe_variable" "var":
  10: resource "tfe_variable" "var" {
```

On the other hand if overwrite is set to true the it will succeed

```
resource "tfe_variable" "var" {
  key          = "my_key"
  value        = "my_value"
  category     = "terraform"
  workspace_id = var.workspace_id
  overwrite = true
}
```

```
OSXSWINKMBP15:tmp swinkler$ terraform apply -auto-approve
tfe_variable.var: Creating...
tfe_variable.var: Creation complete after 1s [id=var-5WZXq3MSnRWpeUAs]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
```
## Output from acceptance tests

fails due to other reasons. tests in `resource_tfe_variable_test.go` all pass.

```
OSXSWINKMBP15:terraform-provider-tfe swinkler$ make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 15m
?   	github.com/terraform-providers/terraform-provider-tfe	[no test files]
=== RUN   TestAccTFESSHKeyDataSource_basic
--- FAIL: TestAccTFESSHKeyDataSource_basic (0.70s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamAccessDataSource_basic
--- FAIL: TestAccTFETeamAccessDataSource_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamDataSource_basic
--- FAIL: TestAccTFETeamDataSource_basic (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspaceIDsDataSource_basic
--- FAIL: TestAccTFEWorkspaceIDsDataSource_basic (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspaceIDsDataSource_wildcard
--- FAIL: TestAccTFEWorkspaceIDsDataSource_wildcard (0.30s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspaceDataSource_basic
--- FAIL: TestAccTFEWorkspaceDataSource_basic (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestProvider_versionConstraints
--- PASS: TestProvider_versionConstraints (0.00s)
=== RUN   TestAccTFENotificationConfiguration_basic
--- FAIL: TestAccTFENotificationConfiguration_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFENotificationConfiguration_update
--- FAIL: TestAccTFENotificationConfiguration_update (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFENotificationConfiguration_slackWithToken
--- FAIL: TestAccTFENotificationConfiguration_slackWithToken (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFENotificationConfiguration_duplicateTriggers
--- FAIL: TestAccTFENotificationConfiguration_duplicateTriggers (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFENotificationConfigurationImport
--- FAIL: TestAccTFENotificationConfigurationImport (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOAuthClient_basic
--- FAIL: TestAccTFEOAuthClient_basic (0.75s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOrganization_basic
--- FAIL: TestAccTFEOrganization_basic (0.12s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOrganization_update
--- FAIL: TestAccTFEOrganization_update (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOrganization_import
--- FAIL: TestAccTFEOrganization_import (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOrganizationToken_basic
--- FAIL: TestAccTFEOrganizationToken_basic (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOrganizationToken_existsWithoutForce
--- FAIL: TestAccTFEOrganizationToken_existsWithoutForce (0.15s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOrganizationToken_existsWithForce
--- FAIL: TestAccTFEOrganizationToken_existsWithForce (1.41s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEOrganizationToken_import
--- FAIL: TestAccTFEOrganizationToken_import (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySetParameter_basic
--- FAIL: TestAccTFEPolicySetParameter_basic (0.34s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySetParameter_update
--- FAIL: TestAccTFEPolicySetParameter_update (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySetParameter_import
--- FAIL: TestAccTFEPolicySetParameter_import (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySet_basic
--- FAIL: TestAccTFEPolicySet_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySet_update
--- FAIL: TestAccTFEPolicySet_update (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySet_updateEmpty
--- FAIL: TestAccTFEPolicySet_updateEmpty (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySet_updatePopulated
--- FAIL: TestAccTFEPolicySet_updatePopulated (0.65s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySet_updateToGlobal
--- FAIL: TestAccTFEPolicySet_updateToGlobal (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySet_updateToWorkspace
--- FAIL: TestAccTFEPolicySet_updateToWorkspace (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySet_vcs
--- FAIL: TestAccTFEPolicySet_vcs (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEPolicySetImport
--- FAIL: TestAccTFEPolicySetImport (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFESentinelPolicy_basic
--- FAIL: TestAccTFESentinelPolicy_basic (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFESentinelPolicy_update
--- FAIL: TestAccTFESentinelPolicy_update (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFESentinelPolicy_import
--- FAIL: TestAccTFESentinelPolicy_import (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFESSHKey_basic
--- FAIL: TestAccTFESSHKey_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFESSHKey_update
--- FAIL: TestAccTFESSHKey_update (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamAccess_basic
--- FAIL: TestAccTFETeamAccess_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamAccess_import
--- FAIL: TestAccTFETeamAccess_import (0.15s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestPackTeamMemberID
--- PASS: TestPackTeamMemberID (0.00s)
=== RUN   TestUnpackTeamMemberID
--- PASS: TestUnpackTeamMemberID (0.00s)
=== RUN   TestAccTFETeamMember_basic
--- FAIL: TestAccTFETeamMember_basic (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamMember_import
--- FAIL: TestAccTFETeamMember_import (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamMembers_basic
--- FAIL: TestAccTFETeamMembers_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamMembers_update
--- FAIL: TestAccTFETeamMembers_update (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamMembers_import
--- FAIL: TestAccTFETeamMembers_import (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeam_basic
--- FAIL: TestAccTFETeam_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeam_import
--- FAIL: TestAccTFETeam_import (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamToken_basic
--- FAIL: TestAccTFETeamToken_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamToken_existsWithoutForce
--- FAIL: TestAccTFETeamToken_existsWithoutForce (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamToken_existsWithForce
--- FAIL: TestAccTFETeamToken_existsWithForce (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFETeamToken_import
--- FAIL: TestAccTFETeamToken_import (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEVariable_basic
--- FAIL: TestAccTFEVariable_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEVariable_update
--- FAIL: TestAccTFEVariable_update (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEVariable_import
--- FAIL: TestAccTFEVariable_import (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestPackWorkspaceID
--- PASS: TestPackWorkspaceID (0.00s)
=== RUN   TestUnpackWorkspaceID
--- PASS: TestUnpackWorkspaceID (0.00s)
=== RUN   TestAccTFEWorkspace_basic
--- FAIL: TestAccTFEWorkspace_basic (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspace_monorepo
--- FAIL: TestAccTFEWorkspace_monorepo (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspace_renamed
--- FAIL: TestAccTFEWorkspace_renamed (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspace_update
--- FAIL: TestAccTFEWorkspace_update (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspace_updateFileTriggers
--- FAIL: TestAccTFEWorkspace_updateFileTriggers (0.10s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspace_sshKey
--- FAIL: TestAccTFEWorkspace_sshKey (0.11s)
    provider_test.go:97: err: required token could not be found
=== RUN   TestAccTFEWorkspace_import
--- FAIL: TestAccTFEWorkspace_import (0.10s)
    provider_test.go:97: err: required token could not be found
FAIL
FAIL	github.com/terraform-providers/terraform-provider-tfe/tfe	11.007s
?   	github.com/terraform-providers/terraform-provider-tfe/version	[no test files]
FAIL
make: *** [testacc] Error 1
OSXSWINKMBP15:terraform-provider-tfe swinkler$ export TFE_TOKEN=ZQpn1rAgXv6SzQ.atlasv1.WisfKzyZ6Zg9j93SHCuoNRNCtOjHcep6z4KFRTCzIt28pUmugO6mwznFzoBpdzUkXa0
OSXSWINKMBP15:terraform-provider-tfe swinkler$ make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 15m
?   	github.com/terraform-providers/terraform-provider-tfe	[no test files]
^Cmake: *** [testacc] Error 1

OSXSWINKMBP15:terraform-provider-tfe swinkler$ export TFE_HOSTNAME=dev.deploy.rd.elliemae.io
OSXSWINKMBP15:terraform-provider-tfe swinkler$ make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 15m
?   	github.com/terraform-providers/terraform-provider-tfe	[no test files]
=== RUN   TestAccTFESSHKeyDataSource_basic
--- FAIL: TestAccTFESSHKeyDataSource_basic (0.92s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamAccessDataSource_basic
--- FAIL: TestAccTFETeamAccessDataSource_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamDataSource_basic
--- FAIL: TestAccTFETeamDataSource_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspaceIDsDataSource_basic
--- FAIL: TestAccTFEWorkspaceIDsDataSource_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspaceIDsDataSource_wildcard
--- FAIL: TestAccTFEWorkspaceIDsDataSource_wildcard (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspaceDataSource_basic
--- FAIL: TestAccTFEWorkspaceDataSource_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestProvider_versionConstraints
--- PASS: TestProvider_versionConstraints (0.00s)
=== RUN   TestAccTFENotificationConfiguration_basic
--- FAIL: TestAccTFENotificationConfiguration_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFENotificationConfiguration_update
--- FAIL: TestAccTFENotificationConfiguration_update (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFENotificationConfiguration_slackWithToken
--- FAIL: TestAccTFENotificationConfiguration_slackWithToken (0.18s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFENotificationConfiguration_duplicateTriggers
--- FAIL: TestAccTFENotificationConfiguration_duplicateTriggers (0.22s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFENotificationConfigurationImport
--- FAIL: TestAccTFENotificationConfigurationImport (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOAuthClient_basic
--- FAIL: TestAccTFEOAuthClient_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOrganization_basic
--- FAIL: TestAccTFEOrganization_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOrganization_update
--- FAIL: TestAccTFEOrganization_update (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOrganization_import
--- FAIL: TestAccTFEOrganization_import (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOrganizationToken_basic
--- FAIL: TestAccTFEOrganizationToken_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOrganizationToken_existsWithoutForce
--- FAIL: TestAccTFEOrganizationToken_existsWithoutForce (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOrganizationToken_existsWithForce
--- FAIL: TestAccTFEOrganizationToken_existsWithForce (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEOrganizationToken_import
--- FAIL: TestAccTFEOrganizationToken_import (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySetParameter_basic
--- FAIL: TestAccTFEPolicySetParameter_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySetParameter_update
--- FAIL: TestAccTFEPolicySetParameter_update (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySetParameter_import
--- FAIL: TestAccTFEPolicySetParameter_import (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySet_basic
--- FAIL: TestAccTFEPolicySet_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySet_update
--- FAIL: TestAccTFEPolicySet_update (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySet_updateEmpty
--- FAIL: TestAccTFEPolicySet_updateEmpty (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySet_updatePopulated
--- FAIL: TestAccTFEPolicySet_updatePopulated (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySet_updateToGlobal
--- FAIL: TestAccTFEPolicySet_updateToGlobal (10.18s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySet_updateToWorkspace
--- FAIL: TestAccTFEPolicySet_updateToWorkspace (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySet_vcs
--- FAIL: TestAccTFEPolicySet_vcs (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEPolicySetImport
--- FAIL: TestAccTFEPolicySetImport (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFESentinelPolicy_basic
--- FAIL: TestAccTFESentinelPolicy_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFESentinelPolicy_update
--- FAIL: TestAccTFESentinelPolicy_update (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFESentinelPolicy_import
--- FAIL: TestAccTFESentinelPolicy_import (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFESSHKey_basic
--- FAIL: TestAccTFESSHKey_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFESSHKey_update
--- FAIL: TestAccTFESSHKey_update (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamAccess_basic
--- FAIL: TestAccTFETeamAccess_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamAccess_import
--- FAIL: TestAccTFETeamAccess_import (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestPackTeamMemberID
--- PASS: TestPackTeamMemberID (0.00s)
=== RUN   TestUnpackTeamMemberID
--- PASS: TestUnpackTeamMemberID (0.00s)
=== RUN   TestAccTFETeamMember_basic
--- FAIL: TestAccTFETeamMember_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamMember_import
--- FAIL: TestAccTFETeamMember_import (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamMembers_basic
--- FAIL: TestAccTFETeamMembers_basic (0.21s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamMembers_update
--- FAIL: TestAccTFETeamMembers_update (0.22s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamMembers_import
--- FAIL: TestAccTFETeamMembers_import (0.19s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeam_basic
--- FAIL: TestAccTFETeam_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeam_import
--- FAIL: TestAccTFETeam_import (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamToken_basic
--- FAIL: TestAccTFETeamToken_basic (0.18s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamToken_existsWithoutForce
--- FAIL: TestAccTFETeamToken_existsWithoutForce (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamToken_existsWithForce
--- FAIL: TestAccTFETeamToken_existsWithForce (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFETeamToken_import
--- FAIL: TestAccTFETeamToken_import (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEVariable_basic
--- FAIL: TestAccTFEVariable_basic (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEVariable_update
--- FAIL: TestAccTFEVariable_update (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEVariable_import
--- FAIL: TestAccTFEVariable_import (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestPackWorkspaceID
--- PASS: TestPackWorkspaceID (0.00s)
=== RUN   TestUnpackWorkspaceID
--- PASS: TestUnpackWorkspaceID (0.00s)
=== RUN   TestAccTFEWorkspace_basic
--- FAIL: TestAccTFEWorkspace_basic (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspace_monorepo
--- FAIL: TestAccTFEWorkspace_monorepo (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspace_renamed
--- FAIL: TestAccTFEWorkspace_renamed (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspace_update
--- FAIL: TestAccTFEWorkspace_update (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspace_updateFileTriggers
--- FAIL: TestAccTFEWorkspace_updateFileTriggers (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspace_sshKey
--- FAIL: TestAccTFEWorkspace_sshKey (0.16s)
    provider_test.go:97: err: host does not provide a tfe service
=== RUN   TestAccTFEWorkspace_import
--- FAIL: TestAccTFEWorkspace_import (0.17s)
    provider_test.go:97: err: host does not provide a tfe service
FAIL
FAIL	github.com/terraform-providers/terraform-provider-tfe/tfe	20.669s
?   	github.com/terraform-providers/terraform-provider-tfe/version	[no test files]
FAIL
make: *** [testacc] Error 1
OSXSWINKMBP15:terraform-provider-tfe swinkler$ export TFE_HOSTNAME=tfe.dev.deploy.rd.elliemae.io
OSXSWINKMBP15:terraform-provider-tfe swinkler$ make testacc
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test $(go list ./... |grep -v 'vendor') -v  -timeout 15m
?   	github.com/terraform-providers/terraform-provider-tfe	[no test files]
=== RUN   TestAccTFESSHKeyDataSource_basic
--- PASS: TestAccTFESSHKeyDataSource_basic (5.16s)
=== RUN   TestAccTFETeamAccessDataSource_basic
--- PASS: TestAccTFETeamAccessDataSource_basic (7.84s)
=== RUN   TestAccTFETeamDataSource_basic
--- PASS: TestAccTFETeamDataSource_basic (4.75s)
=== RUN   TestAccTFEWorkspaceIDsDataSource_basic
--- PASS: TestAccTFEWorkspaceIDsDataSource_basic (5.49s)
=== RUN   TestAccTFEWorkspaceIDsDataSource_wildcard
--- PASS: TestAccTFEWorkspaceIDsDataSource_wildcard (5.82s)
=== RUN   TestAccTFEWorkspaceDataSource_basic
--- PASS: TestAccTFEWorkspaceDataSource_basic (4.66s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestProvider_versionConstraints
--- PASS: TestProvider_versionConstraints (0.00s)
=== RUN   TestAccTFENotificationConfiguration_basic
--- PASS: TestAccTFENotificationConfiguration_basic (4.84s)
=== RUN   TestAccTFENotificationConfiguration_update
--- PASS: TestAccTFENotificationConfiguration_update (7.67s)
=== RUN   TestAccTFENotificationConfiguration_slackWithToken
--- PASS: TestAccTFENotificationConfiguration_slackWithToken (2.76s)
=== RUN   TestAccTFENotificationConfiguration_duplicateTriggers
--- PASS: TestAccTFENotificationConfiguration_duplicateTriggers (4.93s)
=== RUN   TestAccTFENotificationConfigurationImport
--- PASS: TestAccTFENotificationConfigurationImport (5.08s)
=== RUN   TestAccTFEOAuthClient_basic
--- SKIP: TestAccTFEOAuthClient_basic (0.18s)
    resource_tfe_oauth_client_test.go:19: Please set GITHUB_TOKEN to run this test
=== RUN   TestAccTFEOrganization_basic
--- PASS: TestAccTFEOrganization_basic (3.10s)
=== RUN   TestAccTFEOrganization_update
--- PASS: TestAccTFEOrganization_update (4.70s)
=== RUN   TestAccTFEOrganization_import
--- PASS: TestAccTFEOrganization_import (2.89s)
=== RUN   TestAccTFEOrganizationToken_basic
--- PASS: TestAccTFEOrganizationToken_basic (4.03s)
=== RUN   TestAccTFEOrganizationToken_existsWithoutForce
--- PASS: TestAccTFEOrganizationToken_existsWithoutForce (5.06s)
=== RUN   TestAccTFEOrganizationToken_existsWithForce
--- PASS: TestAccTFEOrganizationToken_existsWithForce (6.91s)
=== RUN   TestAccTFEOrganizationToken_import
--- PASS: TestAccTFEOrganizationToken_import (4.11s)
=== RUN   TestAccTFEPolicySetParameter_basic
--- PASS: TestAccTFEPolicySetParameter_basic (6.02s)
=== RUN   TestAccTFEPolicySetParameter_update
--- PASS: TestAccTFEPolicySetParameter_update (9.68s)
=== RUN   TestAccTFEPolicySetParameter_import
--- PASS: TestAccTFEPolicySetParameter_import (6.07s)
=== RUN   TestAccTFEPolicySet_basic
--- PASS: TestAccTFEPolicySet_basic (5.99s)
=== RUN   TestAccTFEPolicySet_update
--- PASS: TestAccTFEPolicySet_update (10.78s)
=== RUN   TestAccTFEPolicySet_updateEmpty
--- PASS: TestAccTFEPolicySet_updateEmpty (7.98s)
=== RUN   TestAccTFEPolicySet_updatePopulated
--- PASS: TestAccTFEPolicySet_updatePopulated (12.06s)
=== RUN   TestAccTFEPolicySet_updateToGlobal
--- PASS: TestAccTFEPolicySet_updateToGlobal (10.20s)
=== RUN   TestAccTFEPolicySet_updateToWorkspace
--- PASS: TestAccTFEPolicySet_updateToWorkspace (10.25s)
=== RUN   TestAccTFEPolicySet_vcs
--- SKIP: TestAccTFEPolicySet_vcs (0.18s)
    resource_tfe_policy_set_test.go:255: Please set GITHUB_TOKEN to run this test
=== RUN   TestAccTFEPolicySetImport
--- PASS: TestAccTFEPolicySetImport (6.11s)
=== RUN   TestAccTFESentinelPolicy_basic
--- PASS: TestAccTFESentinelPolicy_basic (5.26s)
=== RUN   TestAccTFESentinelPolicy_update
--- PASS: TestAccTFESentinelPolicy_update (9.08s)
=== RUN   TestAccTFESentinelPolicy_import
--- PASS: TestAccTFESentinelPolicy_import (5.47s)
=== RUN   TestAccTFESSHKey_basic
--- PASS: TestAccTFESSHKey_basic (4.09s)
=== RUN   TestAccTFESSHKey_update
--- PASS: TestAccTFESSHKey_update (6.36s)
=== RUN   TestAccTFETeamAccess_basic
--- PASS: TestAccTFETeamAccess_basic (6.21s)
=== RUN   TestAccTFETeamAccess_import
--- PASS: TestAccTFETeamAccess_import (5.80s)
=== RUN   TestPackTeamMemberID
--- PASS: TestPackTeamMemberID (0.00s)
=== RUN   TestUnpackTeamMemberID
--- PASS: TestUnpackTeamMemberID (0.00s)
=== RUN   TestAccTFETeamMember_basic
--- FAIL: TestAccTFETeamMember_basic (2.91s)
    testing.go:568: Step 0 error: errors during apply:

        Error: Error adding user "admin" to team team-kbeW9wiTVc1XndUi: bad request

        admin does not exist

          on /var/folders/fc/h7kqx4g50t34rsljrz7nnmg1sk0y_z/T/tf-test759953370/main.tf line 12:
          (source code not available)


=== RUN   TestAccTFETeamMember_import
--- FAIL: TestAccTFETeamMember_import (2.90s)
    testing.go:568: Step 0 error: errors during apply:

        Error: Error adding user "admin" to team team-5iDzHQ3PXkB62FDP: bad request

        admin does not exist

          on /var/folders/fc/h7kqx4g50t34rsljrz7nnmg1sk0y_z/T/tf-test630341660/main.tf line 12:
          (source code not available)


=== RUN   TestAccTFETeamMembers_basic
--- SKIP: TestAccTFETeamMembers_basic (0.19s)
    resource_tfe_team_members_test.go:22: Please set TFE_USER1 to run this test
=== RUN   TestAccTFETeamMembers_update
--- SKIP: TestAccTFETeamMembers_update (0.19s)
    resource_tfe_team_members_test.go:55: Please set TFE_USER1 to run this test
=== RUN   TestAccTFETeamMembers_import
--- SKIP: TestAccTFETeamMembers_import (0.19s)
    resource_tfe_team_members_test.go:102: Please set TFE_USER1 to run this test
=== RUN   TestAccTFETeam_basic
--- PASS: TestAccTFETeam_basic (3.87s)
=== RUN   TestAccTFETeam_import
--- PASS: TestAccTFETeam_import (4.03s)
=== RUN   TestAccTFETeamToken_basic
--- PASS: TestAccTFETeamToken_basic (5.17s)
=== RUN   TestAccTFETeamToken_existsWithoutForce
--- PASS: TestAccTFETeamToken_existsWithoutForce (6.45s)
=== RUN   TestAccTFETeamToken_existsWithForce
--- PASS: TestAccTFETeamToken_existsWithForce (8.78s)
=== RUN   TestAccTFETeamToken_import
--- PASS: TestAccTFETeamToken_import (5.36s)
=== RUN   TestAccTFEVariable_basic
--- PASS: TestAccTFEVariable_basic (6.40s)
=== RUN   TestAccTFEVariable_update
--- PASS: TestAccTFEVariable_update (10.88s)
=== RUN   TestAccTFEVariable_import
--- PASS: TestAccTFEVariable_import (6.49s)
=== RUN   TestPackWorkspaceID
--- PASS: TestPackWorkspaceID (0.00s)
=== RUN   TestUnpackWorkspaceID
--- PASS: TestUnpackWorkspaceID (0.00s)
=== RUN   TestAccTFEWorkspace_basic
--- PASS: TestAccTFEWorkspace_basic (3.89s)
=== RUN   TestAccTFEWorkspace_monorepo
--- PASS: TestAccTFEWorkspace_monorepo (4.15s)
=== RUN   TestAccTFEWorkspace_renamed
--- PASS: TestAccTFEWorkspace_renamed (6.23s)
=== RUN   TestAccTFEWorkspace_update
--- FAIL: TestAccTFEWorkspace_update (7.50s)
    testing.go:568: Step 1 error: Check failed: Check 2/12 error: Bad operations: true
=== RUN   TestAccTFEWorkspace_updateFileTriggers
--- PASS: TestAccTFEWorkspace_updateFileTriggers (6.36s)
=== RUN   TestAccTFEWorkspace_sshKey
--- PASS: TestAccTFEWorkspace_sshKey (9.87s)
=== RUN   TestAccTFEWorkspace_import
--- PASS: TestAccTFEWorkspace_import (4.14s)
FAIL
FAIL	github.com/terraform-providers/terraform-provider-tfe/tfe	323.719s
?   	github.com/terraform-providers/terraform-provider-tfe/version	[no test files]
FAIL

...
```

beekus committed:
Add workspace state updater

beekus committed:
Extract workspace helpers

beekus committed:
Use external ids in workspace resource

beekus committed:
Update external id in workspace data source

beekus committed:
Add fetchHumanId

beekus committed:
Migrate variable

beekus committed:
Migrate Team Access

beekus committed:
Acceptance tests fixup

beekus committed:
Fix the importer for var and team access

beekus committed:
Add validator func for workspace_id

beekus committed:
Use a better regular expression

beekus committed:
Constantize the workspace id regex

beekus committed:
Update team_access.html.markdown

beekus committed:
Update workspace.html.markdown

beekus committed:
Update workspace_ids.html.markdown

beekus committed:
Update team_access.html.markdown

beekus committed:
Update workspace.html.markdown

beekus committed:
Update variable.html.markdown

beekus committed:
Mock the tfeclient for better unit testing

beekus committed:
Require tfe.v2.2

beekus committed:
Merge branch 'master' into f-workspace-id-migration

beekus committed:
Remove some commented code

beekus committed:
Apply suggestions from code review

Co-Authored-By: Sander van Harmelen <sander@vanharmelen.nl>

beekus committed:
Docs fixup

lafentres committed:
Add support for run trigger resource

nicsnet committed:
Upgrade terraform to version 0.12.7 in order to satisfy sdk migration requirements

nicsnet committed:
Migrates to github.com/hashicorp/terraform-plugin-sdk v1.0.0

lafentres committed:
Add tests for run trigger resource

lafentres committed:
Add documentation for run trigger resource

Commit:
add overwrite attribute

Commit:
add overwrite attribute

Commit:
remove go.mod change

Commit:
remove go.mod change

Commit:
ignore variable

Commit:
ignore variable

cam-stitt committed:
Update pull request template to add support for breaking changes

nicsnet committed:
Updates Changelog

nicsnet committed:
Merge pull request #130 from terraform-providers/nicsnet/migrate-to-terraform-sdk

Migrate to new terraform plugin sdk

Commit:
v0.13.0

Commit:
Cleanup after v0.13.0 release

lafentres committed:
Update go-tfe to v0.5.0, run go mod vendor & go mod tidy

cam-stitt committed:
Update README.md

Co-Authored-By: Krista LaFentres <lafentres@users.noreply.github.com>

cam-stitt committed:
Update changelog categories

sheldonhull commented:
Related Issue: #105 
Also partially relevant, due to the lack of ability to easily apply variable updates with overwrite: #104 

I hope this is accepted and appreciate your effort in submitting as it's been a major pain point for me with increased workspace count and desire to better update variables. 
My main solution is currently to use data references to point to credentials so that they don't need to exist on the workspace at all. However, this still has some great use for me as well.

cam-stitt committed:
Merge pull request #134 from terraform-providers/cam-stitt/breaking-change-doc

Update pull request template to add support for breaking changes

lafentres committed:
Merge pull request #132 from terraform-providers/lafentres/run-triggers

Add support for run triggers

lafentres committed:
Update CHANGELOG.md

Commit:
v0.14.0

Commit:
Cleanup after v0.14.0 release

Commit:
Working directory support for updating

Commit:
Fix for test to pass update

Commit:
Added to to change workspace to default

scottwinkler commented:
@sheldonhull I wrote a quick script to delete variables with a certain name from all workspaces in all organizations (except the one named "Admin") to deal with the issue I was having. You might find this helpful while this pull request is pending:

```
package main

import (
	"context"
	"fmt"
	"log"

	tfe "github.com/hashicorp/go-tfe"
)

func main() {
	config := &tfe.Config{
		Token:   "<token>"
		Address: "<address>",
	}

	client, err := tfe.NewClient(config)
	if err != nil {
		log.Fatal(err)
	}

	// Create a context
	ctx := context.Background()

	ol, _ := client.Organizations.List(ctx, tfe.OrganizationListOptions{})
	for _, o := range ol.Items {
		fmt.Printf("Organization: %s\n", o.Name)
		if o.Name == "Admin" {
			fmt.Printf("Skipping Admin Organization\n")
			continue
		}
		wl, _ := client.Workspaces.List(ctx, o.Name, tfe.WorkspaceListOptions{})
		for wl.CurrentPage <= wl.TotalPages {
			for _, w := range wl.Items {
				fmt.Printf("  Workspace: %s, Page: %d\n", w.Name, wl.CurrentPage)
				vl, _ := client.Variables.List(ctx, w.ID, tfe.VariableListOptions{}) 
				for _, v := range vl.Items {
					if v.Key == "AWS_ACCESS_KEY_ID" || v.Key == "AWS_SECRET_ACCESS_KEY" {
						fmt.Printf("    Found Matching Variable: %s\n", v.Key)
						fmt.Printf("    Deleting Matching Variable: %s\n", v.Key)
						client.Variables.Delete(ctx, w.ID, v.ID)
					}
				}
			}
			if wl.CurrentPage == wl.TotalPages {
				break
			}
			wl, _ = client.Workspaces.List(ctx, o.Name, tfe.WorkspaceListOptions{ListOptions: tfe.ListOptions{PageNumber: wl.NextPage}})

		}
	}
}
```

nicsnet committed:
Reset trigger prefixes for workspace resources when they are omitted from the config

nicsnet committed:
Fixes test for updating the workspace

Commit:
Create to use default or supplied working directory config

Commit:
Revert changes for test fix

Commit:
Added dedicated test for working directory update

nicsnet committed:
Merge pull request #137 from etopeter/workspace-working-directory-updates

Workspace working directory updates

nicsnet committed:
Updates Changelog

nicsnet committed:
Add links to PRs for Changelog update

nicsnet committed:
Merge pull request #138 from terraform-providers/nicsnet/bugfix-cleanup-trigger-prefixes-for-workspaces

Reset trigger prefixes for workspace resources when removed from the config

Commit:
v0.14.1

Commit:
Cleanup after v0.14.1 release

lafentres committed:
Remove beta flag from tfe_run_trigger resource documentation

lafentres committed:
Merge pull request #143 from terraform-providers/lafentres/docs-run-triggers-beta-removal

Remove beta flag from tfe_run_trigger resource documentation

beekus committed:
Merge branch 'master' into f-workspace-id-migration

beekus committed:
Update for TF plugin sdk

beekus committed:
Update variable to use workspace read by id

beekus committed:
Update CHANGELOG.md

Commit:
v0.15.0

beekus committed:
Merge branch 'master' into f-workspace-id-migration

beekus committed:
Merge pull request #106 from terraform-providers/f-workspace-id-migration

Workspace id migration

Commit:
v0.15.1

Commit:
Cleanup after v0.15.1 release

lafentres committed:
Add documentation on how to constrain the acceptable provider versions to readme and website docs

lafentres committed:
Merge pull request #148 from terraform-providers/lafentres/provider-version-docs

Add documentation on constraining provider versions

Commit:
Fix panic if workspace no longer exists

Commit:
Add resource_tfe_team_organization_membership

Allows adding a user to a team through the organization membership
resource.

beekus committed:
Document team organization memberships

* Add docs for organization_membership
* Add docs for team_organization_member
* Change notes on team_member and team_members to suggest that
  team_organization_member is the preferred resource for managing
  team membership.

bendrucker committed:
docs: update import ID for tfe_workspace

beekus committed:
Add docs with PTFE version notes

lafentres committed:
Merge pull request #156 from bendrucker/docs-workspace-import-by-id

docs: update import ID for tfe_workspace

lafentres committed:
Add note to tfe_workspace resource documentation explaining the old format of workspace ids.

lafentres committed:
Fix broken code blocks in index page of documentation

lafentres committed:
Add note explaining the difference between ids and external_ids for tfe_workspace_ids data source

Commit:
Use go-tfe v0.6.0

lafentres committed:
Merge pull request #159 from terraform-providers/lafentres/workspace_ids_docs

Update docs to reflect changes to workspace IDs + fix broken code block

sheldonhull commented:
Th- at's super cool! I actually did this in powershell but as I'm trying to learn go this is fantastic example of a go utility. Thanks!

Commit:
Rename 'my-workflow' -> 'build-and-test'

beekus committed:
Add a note about updating email addresses

leetrout committed:
Update CHANGELOG.md

beekus committed:
Merge pull request #154 from terraform-providers/leetrout/use-org-memberships-for-teams

Use org memberships for teams

Commit:
v0.16.0

Commit:
Cleanup after v0.16.0 release

Commit:
Add panicking test

Commit:
Log missing workspace

Example Output:
2020/04/26 15:48:27 [DEBUG] Workspace ws-CJygBQQU156fwXGF no longer exists

Commit:
Update CHANGELOG for 0.16.1

leetrout committed:
Merge pull request #162 from terraform-providers/leetrout/fix-nil-panic

Fix workspace panic

Commit:
v0.16.1

Commit:
Cleanup after v0.16.1 release

Commit:
Merge branch 'feature/var-overwrite' of https://github.com/scottwinkler/terraform-provider-tfe into feature/var-overwrite

scottwinkler commented:
rebased

scottwinkler commented:
this ongoing problem literally costs my team hours of work each week. who can I contact to escalate this?

koikonom commented:
Hi @scottwinkler, can you please take a look at https://github.com/terraform-providers/terraform-provider-tfe/pull/175 and let me know if that helps with your issue or not?

scottwinkler commented:
@koikonom yes, i do think the previous PR is useful but it solves a different problem than this one.

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-provider-tfe?pullRequest=133)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-provider-tfe?pullRequest=133)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**6** out of **7** committers have signed the CLA.
- [x] lafentres
- [x] nicsnet
- [x] cam-stitt
- [x] bendrucker
- [x] beekus
- [x] leetrout
- [ ] Scott Winkler

----

**Scott Winkler** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA.
If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).


<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-provider-tfe?pullRequest=133).</sub>

chrisarcand commented:
Hi @scottwinkler! I'd love to help get this resolved for you. 

So to summarize, if only for my own sake, it looks like this is what happened (correct if I'm wrong!):

* The `tfe_variable` resource was changed in https://github.com/hashicorp/terraform-provider-tfe/commit/52199f016ef0638363deab1b6177492dbb43de6b ...
* ...to use use the new client API in https://github.com/hashicorp/go-tfe/pull/97/commits/3c08ccf4d60964869dba86d89a0eec8c8c92ed98 ...
* ... which utilizes the newer [Workspace Variables API](https://www.terraform.io/docs/cloud/api/workspace-variables.html) ...
* ... which is a breaking change for you updating to v0.12.0 of the provider, running some version of Terraform Enterprise prior to `v202001-1`, as any version prior to that does not yet include that newer API, thus losing your variable resources.

That's a bummer, and I'm sorry installing 0.12.0 broke things for you. Although we do warn in the [README](https://github.com/hashicorp/terraform-provider-tfe/blob/master/README.md) and [provider documentation](https://registry.terraform.io/providers/hashicorp/tfe/latest/docs) that any minor version changes indicate possible breaking changes, this certainly could have been avoided and issue surfaced in a less obtrusive way. :( This oversight was fortunately corrected in https://github.com/hashicorp/terraform-provider-tfe/pull/127 (v0.12.1)

The solution proposed here seems like a bit of a feature hammer for a one-off issue between pre-1.0 releases, and "this fixes an annoying problem where people create variables in the UI with the same key name (for example for access keys) which then causes my Terraform configuration to fail." contradicts the notion of letting Terraform manage your resources and not also combine out-of-band changes to those resources at the same time. Terraform is to be considered your single source of truth, and not just continually overwrite over the top of any such out-of-band changes.

Instead of going this route with this `overwrite` attribute, I propose one of two paths forward:

* **Upgrade TFE before moving to 0.12.x+.** Since it's been so long, I should ask: do you still have 1,000 variable resources unmanaged in your Terraform state? If you've upgraded TFE since the issue arose and are not still on 0.11.x, the state _should_ correct itself using the new API as it does not appear that the identifier in state has changed - it would just start to use the new API instead with the same identifier. If you've already written to your state without a first creating a backup, you also have the option of making a custom build with this `overwrite` change just to correct this one-off issue and migrate easily. Anyone else moving to 0.12.x at this point would avoid this issue with the compatibility check now in place.
* **If a TFE upgrade is not an option, write a more graceful API transition.** The [Variables API](https://www.terraform.io/docs/cloud/api/variables.html) that was being used in versions prior to 0.12.0 is now long deprecated, but if for some reason you are even now unable to upgrade TFE to `v202001-1` (which again, includes the new Workspace Variables API), what I would do is submit a change to go-tfe and this provider to support the deprecated API again as well. The older API could be attempted if the variable cannot be found using the new one. This should allow variables to be properly read using old versions of TFE (for now!) and solve your issue. This is what would have had to have been done with a post-1.0 provider without making a major release.

It's unfortunate that 0.12.0 caused issues; we try hard to maintain backwards compatibility even as we're still in the 0.x release series and at least try and mark when something isn't backwards compatible (as was done in 0.12.1), but missed it there. Let me know where you're at with the unmanaged variables and we'll get this issue closed out 👍🏻

Thanks Scott!

chrisarcand commented:
Closing due to inactivity; but absolutely feel free to reach out if you're still having trouble with the unmanaged variables, as I said above. Thanks!

scottwinkler commented:
i would like to reopen this. its not having anything to do with Terraform 0.12 compatability. This is a problem in the latest version of Terraform. The thing is that I want a way to set a variable in a workspace and overwrite any existing variable with the same name that might live in that workspace. I think this is a straightforward ask, and I have definetly seen this one other Terraform providers. In one case i saw it was called "force_overwrite"

chrisarcand commented:
It is technically straightforward, but ill-advised as it is 'forcing' your Terraform configuration to be the only source of truth when it is unfortunately not. The reality is that any out-of-band changes (which cannot be controlled by Terraform) contradict it.

The precedence you're looking for can be found in something like the AWS provider's `route53_record` resource, which has an [`allow_overwrite` argument](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_record#allow_overwrite) doing exactly what you're looking for:

![image](https://user-images.githubusercontent.com/2430490/100670286-425a8580-3324-11eb-83b4-899704aec8cf.png)

 _However_, the reason this argument exists is the complete opposite situation - this resource, in 1.x of the provider, overwrote any existing values. This argument was [added as an escape hatch to keep compatibility in 2.x of the provider](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-2-upgrade#allow_overwrite-default-value-change), where properly importing your records is advised:

![image](https://user-images.githubusercontent.com/2430490/100669628-4afe8c00-3323-11eb-9670-8a62b1934574.png)

I think adding this option for us here would be moving backwards, then - but feel free to open a fresh issue with this request to discuss further and get other opinions, of course. (I would not re-use this PR as it's filled with now-unrelated concerns about the breaking provider version, and the change history has diverted very far from `master` at this point)

scottwinkler commented:
yes i understand that this is counter to how Terraform normally works. But i think an exception should be made in this case because this isn't a regular Terraform resource. We aren't talking about an EC2 instance or a load balancer. This is a variable on a workspace. So the only use case of this resource is to keep variables on workspaces in a certain state. How is is that I can prevent my users from replacing certain variables with their own? impossible without removing their access to put any other variables on a workspace. instead i have to resolve a pointless error message instead of just overwriting whatever variable they put on their workspace. if there was a way to prevent users from overwriting this themselves then i would agree with you but the user access policies are not so granular as you are suggesting

sheldonhull commented:
Based on @scottwinkler message, I wanted to also add my follow-up from the conversation. I have monitored this thread and while I completely see the value from @chrisarcand details about the "source of truth", I also didn't agree with the conclusion for *variables*. I completely agree that Terraform's model of being the source of truth makes perfect sense. However, in the context of a variable, this seems to fall into something more like a "configuration" value than an infrastructure component. The easier it is to ensure the values are set the easier it is to adopt using Terraform to manage workspaces. 

I'd suggest that the idea of allowing overwrite of variables for a workspace would prioritize the user experience and helping folks adapt to terraform managed workspaces without lots of manual workarounds or more manual API work.

I don't manage all my workspaces with Terraform currently, but based on my last experience, the pain of dealing with variables being set resulted in me finding different ways to configure my workspaces to avoid them in the first place as it was that disruptive.

Regardless, appreciate the follow-ups and insight into the decision making process. Cheers 👍

chrisarcand commented:
Ahh see, this is why I said I'd be happy to continue discussing!  🤗  

This seems to be well understood, but for posterity: my point was that preventing other users from replacing certain values or resources is impossible, and true of _anything_ in Terraform, given out-of-band changes. The initial reasoning for this feature seemed to be an attempt to do just that. 

My early concern - and where I should have prompted more discussion earlier and I'm happy we're discussing now - was that while I understand the usefulness, a line needs to be clearly drawn about what is acceptable to do this with (and not!). How is a variable _really_ different from any other resource? A provider which simply overwrites all resources is not a very useful thing :) The distinction that was attempted to be made above here wasn't clear to me until @sheldonhull defined it well:

> I completely agree that Terraform's model of being the source of truth makes perfect sense. However, in the context of a variable, this seems to fall into something more like a "configuration" value than an infrastructure component. 

This makes sense and better defines that figurative line I'm talking about. As a variable is nothing more than a key and value associated with a particular workspace, it really could just be a configuration value; and if it were, drift would be properly detected and changed as expected. This is very fair and solid reasoning.

Further revisiting this, there's another aspect which I hadn't thought of before that lends itself well to allowing this behavior: In SDK V2 of the provider, which [this provider should be using shortly](https://github.com/hashicorp/terraform-provider-tfe/pull/246), a [diagnostic API is available](https://www.terraform.io/docs/extend/guides/v2-upgrade-guide.html#support-for-diagnostics) via contexts to supply warnings to Terraform, even on specific arguments. The behavior defined here, which is known to be counter to how Terraform normally works, is then very reasonable. The user who opted in to the possibility of an overwrite can also be warned when it actually occurs.

---

Given all of this, I hear you, this seems reasonable to add. An opt-in mechanism which emits warnings of the atypical behavior being used for a resource - which is uniquely challenging to deal with because it exists more as a configuration value for another resource rather than a normal one simply associated with another - is a well formed distinction. 

@scottwinkler If you (or anyone else) would like to re-open this pull request (or open another) and rebase, adding tests and documentation, I'd be happy to merge it. I do prefer the name `allow_overwrite` as well. I also expect that the SDK upgrade will be merged sooner, so the diagnostic concerns I mentioned above could be added as well.

Cheers indeed, I also very much appreciate the responses and am happy we were able to sort it. 👍🏻

scottwinkler commented:
Sure I would be happy to do this. Please notify me when the v2 sdk update is merged and I will work off of that
